### PR TITLE
Issue #318: Enhance traverse message when no action is done

### LINKED ISSUE #318: git machete traverse didn't work
# Overview

- Working on this repo.
- Made more commits on feature branch.
- `git push` new commits to feature branch
- `git fetch`- develop behind upstream
- I'm on feature branch, execute `git machete traverse` - "nothing left to update"
- `git checkout master`, then `git machete traverse` - performs necessary updates

## Expected behavior/command output

I'm not sure if this is as designed, but I've expected for `git machete traverse` to work the same regardless what is checked out. `git machete traverse --help`:

> Traverses the branch dependency tree in pre-order (i.e. simply in the order as they occur in the definition file) and for each branch

## Actual behavior/command output

image

## `git` and `git machete` versions

```
git version 2.33.1
git machete version 3.4.1
```

### PR DESCRIPTION
**Pull request by rifel123**
closes #318

### CODE REVIEW

File: git_machete/client.py [hunk 1]
```diff
@@ -843,7 +847,8 @@ def traverse(
         else:
             msg = f"No successor of {bold(current_branch)} needs to be slid out or synced with upstream branch or remote"
         sys.stdout.write(f"{msg}; nothing left to update\n")
-
+        if not is_action_done and initial_branch != self.managed_branches[0]:
+            sys.stdout.write(fmt("Tip: `traverse` by default starts from the current branch, use flags (--starts-from=, --whole or -w, -W) to change this behavior.\nFurther info under `git machete traverse --help`.\n"))
```
Comment (PawelLipski) [inline git_machete/client.py]: Pls split the line + surround options with backticks

File: git_machete/client.py [hunk 2]
```diff
@@ -843,7 +847,8 @@ def traverse(
         else:
             msg = f"No successor of {bold(current_branch)} needs to be slid out or synced with upstream branch or remote"
         sys.stdout.write(f"{msg}; nothing left to update\n")
-
+        if not is_action_done and initial_branch != self.managed_branches[0]:
```
Comment (PawelLipski) [inline git_machete/client.py]: Actually, as I'm now looking into the original issue #318... AFAICS this message would still NOT be printed in Konrad's case, right? maybe the just the `not is_action_done` part should be left (?)

File: git_machete/client.py [hunk 3]
```diff
@@ -636,6 +636,7 @@ def traverse(
         self.expect_at_least_one_managed_branch()
 
         self.__empty_line_status = True
+        is_action_done: bool = False
```
Comment (PawelLipski) [inline git_machete/client.py]: `any_action_done`? also, this name is not adequate... rather `any_action_suggested` (?)

File: git_machete/client.py (hunk 2 above)
Comment (rifel123) [reply git_machete/client.py]: It depends on case, there are several in Konrad's example:
- start from `issue/290-fish-completion` where there is nothing to be done: Tip will be printed.
- start from `develop` where some action actions needs to be done: Tip will not be printed.

And other possible cases:
- start from `develop` but everything is synced: Tip will be printed.
- start from `master` : Tip will not ever be printed, cause we checked all branches from the definition file.

Comment (rifel123) [reply git_machete/client.py]: hmm, maybe it should not print tip when starts from any of root branch?

File: docs/source/cli_help/traverse.rst [hunk 4]
```diff
@@ -18,7 +18,9 @@ traverse
                          [--[no-]push] [--[no-]push-untracked]
                          [--return-to=WHERE] [--start-from=WHERE] [-w|--whole] [-W] [-y|--yes]
 
-Traverses the branch dependency tree in pre-order (i.e. simply in the order as they occur in the definition file) and for each branch:
+Traverses the branch dependency tree in pre-order (i.e. simply in the order as they occur in the definition file).
+By default traverse starts from current branch, however this behaviour can be customized using options: --start-from=, --whole or -w, -W.
```
Comment (PawelLipski) [inline docs/source/cli_help/traverse.rst]: Backticks pls (double backticks, in fact)

File: git_machete/docs.py [hunk 5]
```diff
@@ -611,7 +611,9 @@
                                        [--[no-]push] [--[no-]push-untracked]
                                        [--return-to=WHERE] [--start-from=WHERE] [-w|--whole] [-W] [-y|--yes]</b>
 
-        Traverses the branch dependency tree in pre-order (i.e. simply in the order as they occur in the definition file) and for each branch:
+        Traverses the branch dependency tree in pre-order (i.e. simply in the order as they occur in the definition file).
+        By default traverse starts from current branch, however this behaviour can be customized using options: --starts-from=, --whole or -w, -W.
```
Comment (PawelLipski) [inline git_machete/docs.py]: Backticks pls

File: git_machete/docs.py [hunk 6]
```diff
@@ -611,7 +611,9 @@
...
+        Command for each branch:
```
Comment (PawelLipski) [inline git_machete/docs.py]:
```suggestion
        For each branch, the command:
```

Comment (PawelLipski) [reply git_machete/docs.py]: Here and in rst

File: git_machete/client.py (hunk 2 above)
Comment (PawelLipski) [reply git_machete/client.py]: Okay! let's try _not print tip when starts from any root branch (not just the **first** root branch)_ approach then

**Review by PawelLipski — APPROVED** [review]

**Pull request merged**